driver/src/price_estimation/dexag/api.rs

```diff
@@ -0,0 +1,167 @@
+#![allow(dead_code)]
+
+use anyhow::{anyhow, Context, Result};
+use ethcontract::Address;
+use isahc::prelude::*;
+use serde::{de::Error as _, Deserialize, Deserializer};
+use serde_with::rust::display_fromstr;
+use std::time::Duration;
+use url::Url;
+
+#[cfg_attr(test, mockall::automock)]
+/// Parts of the dex.ag api.
+pub trait DexagApi {
+    /// https://docs.dex.ag/api/tokens
+    fn get_token_list(&self) -> Result<Vec<Token>>;
+    /// https://docs.dex.ag/api/price
+    fn get_price(&self, from: &Token, to: &Token, amount: EitherAmount) -> Result<Price>;
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct Token {
+    pub name: String,
+    pub symbol: String,
+    #[serde(default)]
+    #[serde(deserialize_with = "deserialize_address")]
+    pub address: Option<Address>,
```

Comment:
Nit: `Address` should implement `Deserialize` as far as I know so I think this shouldn’t be needed.

Reply:
Problem is that the default implementation does not allow the Strings to start with `0x`.

Reply:
AFAIK that is only true of the `FromStr` implementation. The `Deserialize` implementations require the address string to start with `”0x”`.

Reply:
You're right.